### DISCUSSION TITLE
Referencing the PAGER variable

### DISCUSSION THREAD

**Discussion by diazona** [bugen/pypipe#24] [General]
I noticed that commit 8fc4b47bf9eb9dd6d0614f9788395ad80c6e4a23 changed the code not to use the value of the `PAGER` environment variable anymore. I'm curious, why make that change, and would you be open to reverting it? `PAGER` is the standard environment variable that people set on many UNIX-like systems to tell programs that need to use a pager which one they should use. So changing pypipe to ignore it seems like a step backwards in terms of compatibility.

**Comment by bugen** [comment]
I had quite a dilemma regarding this matter. Since the PAGER environment variable is widely used in various contexts, it is often set for purposes other than changing pypipe's Pager. I decided not to reference the PAGER environment variable to avoid changing pypipe's Pager when the user sets PAGER for purposes other than pypipe. It is possible to revert to referencing PAGER. However, I am considering automatically adding the -R -F -K options when less is specified in PAGER. What are your thoughts on this?

I also struggled with whether making automatic paging the default behavior is a good idea. Personally, I find it convenient, but I understand that some users may dislike such interventions. It's challenging to balance between making the default behavior something that most users would prefer and allowing customization through settings.

**Reply by diazona** [reply]
Thanks for replying and sharing your thoughts. These are interesting questions to consider.

Personally, when I have questions like this with my own software, I try to look at what other tools with similar purposes do. I think it helps the user most when different programs in the same software ecosystem behave consistently, so if I copy other tools, I'm likely to be consistent with what users expect. And probably the most common program which uses a pager is `man`, which does this:

> **-P** _pager_, **--pager=**_pager_
>
> Specify which output pager to use.  By default, `man` uses `less`, falling back to `cat` if `less` is not found or is not executable. This option overrides the `$MANPAGER` environment variable, which in turn overrides the `$PAGER` environment variable. [...]

_If_ you are going to use a pager at all, I think a similar strategy makes sense here:

- Have a command-line option that can be used to set the pager
- If the CLI option is not set, check for a pypipe-specific environment variable (`PYPIPE_PAGER`, I suppose)
- If neither the CLI option nor the pypipe-specific variable are set, use `PAGER`
- If no pager can be determined from any of those, use `less`
- Or as a last resort, use `cat`

The reason I think it's important to reference `PAGER` is that, as a user, when I set `PAGER` I want _every_ program that uses a pager to use that value. For example, imagine that I have [some fancy pager](https://jedsoft.org/most/) on my system that I think is better than `less`. I would install that pager and set `PAGER` so that applications use it, and I'd get used to seeing man pages and such things with all the fancy features my new pager provides. Then I try running something through `pypipe` and the output is nothing like what I expected: no coloring, no multiple window support, most of my familiar key bindings don't work, and so on. I would find that _very_ annoying.

In the rare cases where someone wants to set a pager for every other program but doesn't want pypipe to use it, having a pypipe-specific variable and a command-line option gives them the flexibility to do that. But I don't think it's likely that someone would want that behavior by default. (I suppose I can't speak for _everybody_ in the world - there are many people who [use standard things in unusual ways](https://xkcd.com/1172/) - but I think I'm familiar with general trends.)

By the way, I'd be a little wary of adding options to `less` if you find that in the `PAGER` variable. I suppose if it really seems to improve the user experience, it wouldn't be the worst thing, but in many cases users will have the `LESS` environment variable set with the options that makes `less` behave the way they want to, and it would be disruptive for pypipe to override that. (Note that this variable is processed by `less` itself; you don't have to, and shouldn't, do anything with it in pypipe.) If `LESS` is unset, then yeah, I guess maybe you could add the options... but then, what if someone really wants to use `less` with no options as their pager? It's possible. And I don't think it should be made difficult for them to do so.

Now, then, there is the question of whether to use a pager at all. Again I would take a cue from other similar tools that are standard, but in this case I would look at text processing tools like `sed` and `awk` and [jq](https://jqlang.github.io/jq/) and [miller](https://miller.readthedocs.io/en/6.9.0/), which almost uniformly do _not_ use a pager themselves. They rely on the user to pipe the output into a pager if desired, which is consistent with [the UNIX philosophy](https://en.wikipedia.org/wiki/Unix_philosophy). Though in this case I think it's not _so_ bad to break with the convention and use a pager (which e.g. `git` does), because it's fairly common that if the output device is a TTY then the user actually _does_ want the output to be paged, and if they really don't then they can always pipe it into `cat`... but that is not the most well known trick, and I'm sure using a pager by default will frustrate some people who might have a hard time thinking about how to get around it. Again, this goes back to the approach of trying to do the thing which is most conventional and least confusing for the most people.

Personally, I am a big fan of the UNIX philosophy and so I would have taken the approach that if you want the output of the program to be paged, you should pipe it into the pager yourself. So if it were me, I would not have had pypipe use a pager at all. But of course, you can take your own approach, and there's also the fact that, now that pypipe has been released and has pager invocation built-in, you may not want to remove it to avoid confusing existing users. (Given that it has not yet reached version 1.0, I think they would figure out how to deal with it, but who knows :shrug:)

If you decide that you do want to keep a pager in the program, I would suggest something like the following:

- Have a CLI option `--pager` which takes one argument, a command to use as the pager, and an option `--no-pager` which  specifies that no pager at all will be used. If either of these options is given, then that determines which pager (if any) gets used, period, and no environment variables or built-in defaults will be referenced.
- Otherwise, if the environment variable `PYPIPE_PAGER` is set, use it as the pager command.
- Otherwise, if `PAGER` is set, use it as the pager command.
- Otherwise, use `less`. In this case and (I suggest) only this case, you can check the `LESS` environment variable, and if it is unset, you can pass the `-R -F -K` options to `less`.
- Otherwise, do not use a pager. (Equivalent to using `cat`.)

Maybe this isn't quite perfect, but I think it comes fairly close to what will be the least surprising to the most people. And you can collect some feedback over time and figure out whether it should be tweaked in the future.

**Comment by bugen** [comment]
Thank you for your feedback. It was very helpful.

What bothered me the most was the Unix philosophy. I also thought that automatically connecting to pager, which I'll just call 'paging', is against Unix philosophy, and I also knew that commands such as jq didn't support paging. However, there are some benefits to supporting paging.

Let's consider jq as an example. To pipe the colored output of jq to less, you would use the following command:
```sh
jq . sample.json -C | less -R
```
Both the `-C` option for jq and the `-R` option for less are necessary. Since `-R` is not enabled by default in less, and jq cannot determine the command it is piping to, the output defaults to no colors (unless the user forces it with `-C`). This could be a bit challenging for beginners. If jq supported paging, it could determine whether to output in color based on the specified pager command and automatically add the `-R` option to less. This would be beneficial for beginners. Another advantage is that when the output is long, forgetting to pipe to a pager and flooding the terminal buffer can be avoided (which is crucial for me).

While this goes against the Unix philosophy, considering these benefits, I decided to keep the paging feature in pypipe. The issue is whether to enable or disable paging by default. Currently, it is enabled by default, but it might be better to have it disabled by default.

> - Have a CLI option --pager which takes one argument, a command to use as the pager, and an option --no-pager which specifies that no pager at all will be used. If either of these options is given, then that determines which pager (if any) gets used, period, and no environment variables or built-in defaults will be referenced.

OK, I plan to add the suggested `--pager` and `--no-pager` options. However, I'm thinking of making the `--pager` option without arguments simply enable paging, as specifying a command here would be equivalent to a straightforward pipe to the specified pager.

> - Otherwise, if the environment variable PYPIPE_PAGER is set, use it as the pager command.

This is already implemented.

> - Otherwise, if PAGER is set, use it as the pager command.

I will revert to referencing PAGER.

> - Otherwise, use less. In this case and (I suggest) only this case, you can check the LESS environment variable, and if it is unset, you can pass the -R -F -K options to less.

Thank you for pointing this out. I overlooked the `LESS` environment variable. However, even if the LESS is set, it might be a good to add the `-R -F -K` options proactively. Especially, the `-R` and `-K` options are crucial, as they are necessary for interpreting ANSI color codes and ensuring stable operation. If users did not appreciate this behavior, they can simply disable paging and connect as usual via pipes to their preferred pager command.

> Otherwise, do not use a pager. (Equivalent to using cat.)

I will implemet so.

P.S I wasn't aware of `most`. I will give it a try 👍

**Reply by diazona** [reply]
That makes sense, in general.

You certainly have a point that using no paging at all is trickier for beginners even though it is what the UNIX philosophy would suggest. (Come to think of it, I have sometimes defined a shell function `jql() { jq -C "$@" | less -R; }` because of the inconvenience of giving those options every time - but then I discovered [jless](https://jless.io/) and [jqp](https://github.com/noahgorstein/jqp).) I do think there is an argument for consistency - that is, since beginners will have to deal with many other commands that never run pagers themselves, if you make pypipe use a pager then it makes it harder because they have to remember which programs use pagers and which don't, even though it is easier when running pypipe. Anyway, I think your plan is totally reasonable. And for the reasons you mentioned, I think it would be a good idea to have paging enabled by default.

I did think of a couple more things that might be useful:

- Make sure pypipe checks whether it's outputting to a terminal (a TTY) or to a pipe, if you're not already doing so. And if it's not outputting to a terminal, then I would suggest it should disable paging and also disable color output unless the appropriate command-line options are given (`--pager` and, say, `--color` - you could add that option).
- It might also be handy to have a configuration file so that users can effectively configure their own default choice of whether to have paging enabled. Say, `~/.config/pypipe.cfg` on UNIX-like systems, and the equivalent on other systems.

So that would suggest the following logic, modified from what was in my previous post:

1. Enable or disable paging (and color) according to the command-line options
2. Otherwise, check whether output is being piped to another program and if not, disable paging (and color)
3. Otherwise, check the config file and enable or disable paging (and color) according to that
4. Otherwise, check `PYPIPE_PAGER`
5. Otherwise, check `PAGER`
6. Otherwise, use `less` with appropriate options as pager
7. If `less` is not available, then disable paging

> OK, I plan to add the suggested `--pager` and `--no-pager` options. However, I'm thinking of making the `--pager` option without arguments simply enable paging, as specifying a command here would be equivalent to a straightforward pipe to the specified pager.

Sounds reasonable. In that case there might be something to be said for naming the option `--paginate` instead, for consistency - like git does. Otherwise it might get confusing, people would have to remember that `man`'s `--pager` option takes an argument but pypipe's does not. Then again, it does make sense that `--pager` and `--no-pager` would be opposites :shrug: 

> However, even if the LESS is set, it might be a good to add the -R -F -K options proactively. Especially, the -R and -K options are crucial, as they are necessary for interpreting ANSI color codes and ensuring stable operation. If users did not appreciate this behavior, they can simply disable paging and connect as usual via pipes to their preferred pager command.

I suppose that makes some sense for `-R`. However, I personally wouldn't want `-K` set; it would make it confusing for me to navigate the UI of `less` because I'm not used to it behaving in that way. In general I would be really wary of making assumptions about which options users want. Personally I have `LESS` set to `-R -M --shift 5` and I generally would not want programs messing with that.

I'm not entirely sure what makes the most sense here, but I would like it if there's at least a way to disable any customization of `less`'s options that pypipe may do, perhaps using a configuration file if you choose to add that.

**Comment by bugen** [comment]
@diazona 

I have made modifications to the paging behavior. 6b88e1b428d0794e8951ec4b65fcebe5c28aa7fa
The modifications include the following:

- Added the `--paging` and `--no-paging` options.
- Revised to again reference the `PAGER` environment variable.
- Modified to disable paging if the specified Pager command does not exist.
- Modified to automatically add the `-R -F` options when `less` is specified as the Pager.
- Modified to allow changing the options added to less through the `PYPIPE_LESS_OPTS` environment variable.

I received your opinion that adding options to less automatically when the `LESS` environment variable is set by the user may not be appropriate. However, I have decided to add options to less regardless of the presence of the `LESS` environment variable. This is because I want to force the inclusion of the `-R` option to interpret ASCII color codes. The options to be added to less can be changed with the `PYPIPE_LESS_OPTS` environment variable. Setting it to an empty string will disable the addition of options.

> Make sure pypipe checks whether it's outputting to a terminal (a TTY) or to a pipe, if you're not already doing so. And if it's not outputting to a terminal, then I would suggest it should disable paging and also disable color output unless the appropriate command-line options are given (--pager and, say, --color - you could add that option).

Paging and color are only enabled when the output is a tty or when using an internal pager. This has been the behavior from before.

To be honest, I regret enabling paging by default. I currently believe it is better to have paging disabled by default, allowing users to enable it at their discretion. I plan to disable paging by default in the next release.

**Reply by diazona** [reply]
Makes sense, thanks for explaining! It's a tricky thing to figure out. 

I think disabling paging by default will improve usability for people who are used to using many other UNIX command line tools, so in that sense it's a very reasonable approach to take :+1:
